drivers/usb/device/usb_dc_sam0.c

```diff
@@ -288,6 +288,24 @@ int usb_dc_attach(void)
 	return 0;
 }
 
+static int usb_dc_release_buffers(void)
+{
+	struct usb_sam0_data *data = usb_sam0_get_data();
+	UsbDeviceDescBank *bank;
+	void *buf;
+
+	/* release the buffers */
+	for (int i = 0; i < ARRAY_SIZE(data->descriptors); i++) {
+		for (int j = 0; j < ARRAY_SIZE(data->descriptors[0].DeviceDescBank); j++) {
+			bank = &data->descriptors[i].DeviceDescBank[j];
+			buf = (void *)bank->ADDR.reg;
```

Comment:
Could `bank->ADDR.reg` be `(uintptr_t)NULL`?

Reply:
Do you mean like this:
```
static int usb_dc_release_buffers(void)
{
	struct usb_sam0_data *data = usb_sam0_get_data();
	UsbDeviceDescBank *bank;
	uintptr_t buf;

	/* release the buffers */
	for (int i = 0; i < ARRAY_SIZE(data->descriptors); i++) {
		for (int j = 0; j < ARRAY_SIZE(data->descriptors[0].DeviceDescBank); j++) {
			bank = &data->descriptors[i].DeviceDescBank[j];
			buf = bank->ADDR.reg;
			k_free(buf);
			bank->ADDR.reg = (uintptr_t)NULL;
		}
	}

}
```

Reply:
Sorry, I mean could it be that the buffer is not allocated for a descriptor? `bank->ADDR.reg = 0; `

Reply:
I do not think so.  This is only place I could find where the driver allocates memory (described originally in issue #49177):
https://github.com/zephyrproject-rtos/zephyr/blob/120df9aa1be6ff41be775646ca789f0e8e5ef5d3/drivers/usb/device/usb_dc_sam0.c#L400-L412

So, it appears these memory allocations are descriptors.
 
If the driver is modified in the future to allocate memory for some other purpose, that could be a problem.  But then again, when the driver is detached and then attached again the interface goes through a complete reset.  So, it would likely be appropriate to deallocate everything at that point anyway.

Reply:
Anything else?

Reply:
How I see it, memory is allocated for a endpoint inside usb_dc_ep_configure(), but in usb_dc_release_buffers() your iterate over an array of endpoint descriptors and call k_free() without to check if it was allocated at all. And even if k_free() checks for NULL, there should be a check (or at least comment) before k_free() since the code can change, like in the line 402 in your comment above. My short question that expressed the same was: "Could bank->ADDR.reg be (uintptr_t)NULL?"

Reply:
That is a good point.  I see now that if k_free() changes down the road and does not check for NULL then it would break this.  So, I will explicitly check for NULL before k_free().  I will also add a comment here and another in usb_dc_ep_configure() where the memory is allocated so any future changes to this driver might consider this issue.

Reply:
Your point also brings up another question I had considered but elected to not pursue because it touches other code.  That is, the memory should likely be deallocated when the ep is disabled in usb_dc_ep_disable().  The problem with this approach Is that the call to usb_disable() (subsys/usb/usb_device.c)  is asynchronous with the usb stack events and the stack will never receive an event when usb_disable() is called.  One way to address this would be to insert a call to usb_dc_ep_disable() from usb_disable() or to generate a DC_RESET event from usb_disable().

For example:

```
int usb_disable(void)
{
	int ret;

	if (usb_dev.enabled != true) {
		/*Already disabled*/
		return 0;
	}

	forward_status_cb(USB_DC_RESET, NULL);

	ret = usb_dc_detach();
	if (ret < 0) {
		return ret;
	}

	/* Disable VBUS if needed */
	usb_vbus_set(false);

	usb_dev.enabled = false;

	return 0;
}
```

and release the memory in usb_dc_ep_disable() where we can iterate over the endpoints.  Thoughts?

Reply:
Further to my comment above about using the DC_RESET event, here is what usb_dc_ep_disable() would look like:
```
int usb_dc_ep_disable(uint8_t ep)
{
+	struct usb_sam0_data *data = usb_sam0_get_data();
	UsbDevice *regs = &REGS->DEVICE;
	uint8_t ep_idx = USB_EP_GET_IDX(ep);
	UsbDeviceEndpoint *endpoint = &regs->DeviceEndpoint[ep_idx];
+	UsbDeviceDescriptor *desc = &data->descriptors[ep_idx];
+	UsbDeviceDescBank *bank;
+	void *buf;

	if (ep_idx >= USB_NUM_ENDPOINTS) {
		LOG_ERR("endpoint index/address out of range");
		return -EINVAL;
	}

	endpoint->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_TRCPT0
				 | USB_DEVICE_EPINTENCLR_TRCPT1
				 | USB_DEVICE_EPINTENCLR_RXSTP;

+	if (USB_EP_DIR_IS_IN(ep)) {
+		bank = &desc->DeviceDescBank[1];
+	} else {
+		bank = &desc->DeviceDescBank[0];
+	}
+
+	buf = (void *)bank->ADDR.reg;
+	/*
+	 * We free the ep descriptor memory that was
+	 * allocated in usb_dc_ep_configure()
+	 */
+	if (buf != NULL) {
+		k_free(buf);
+		bank->ADDR.reg = (uintptr_t) NULL;
+	}

	return 0;
}
```

Reply:
The DC_RESET event method will work at the moment because all existing calls in usb_device.c to usb_dc_ep_enable() also have a call to usb_dc_ep_configure().  However, this breaks if usb_dc_ep_enable() is called without usb_dc_ep_configure().  Strictly speaking, the description for usb_dc_ep_enable() in usb_dc.h is silent about usb_dc_ep_configure().  That would need to be changed to indicate that configure is required anytime an ep is disabled.

Reply:
I cursory look at the other drivers and there seems to be several that do not check if the ep is already configured.  The set_endpoint() function in usb_device.c checks the return code of ep_configure() so there are likely issues with other drivers regarding the configure / enable / disable process.

Reply:
IIRC we have an issue open about that, however reconfiguration should not be a big issue, but it should always be checked if an endpoint is already enabled or not.

> One way to address this would be to insert a call to usb_dc_ep_disable() from usb_disable() or to generate a DC_RESET event from usb_disable().

I think usb_disable() is neglected because it is rarely used. There is [foreach_ep(disable_interface_ep); ](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/usb_device.c#L1191), you could add it to usb_disable() just after usb_dc_detach().

Reply:
> There is [foreach_ep(disable_interface_ep); ](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/usb_device.c#L1191), you could add it to usb_disable() just after usb_dc_detach().

IIRC, I think there is a scope issue and the foreach_ep macro cannot be used from usb_device.c.  I think the DC_RESET approach avoids having to change the scope of some of the driver functions and the api.

Reply:
Please follow the link, foreach_ep() is local funciton of usb_device.c

Reply:
okay